ick/runner.py

```diff
@@ -59,33 +62,71 @@ def iter_rule_impl(self) -> Iterable[BaseRule]:
+        """
+        with ThreadPoolExecutor() as tpe:
+            print("[dim]testing...[/dim]")
+            # It's about this point I realize that yes, I am writing a whole
+            # test runner and that's not the business I want to be in.  Oh
+            # well...
+            buffered_output = []
+
+            i = 0
             for rule_instance, names in self.iter_tests():
+                outstanding = {}
+                print(f"  [bold]{rule_instance.rule_config.qualname}[/bold]: ", end="")
                 rule_instance.prepare()
-                progress.update(prepare_key)
                 if not names:
-                    progress.console.print("no tests for", rule_instance.rule_config.qualname, "under", rule_instance.rule_config.test_path)
+                    print("<no-test>", end="")
+                    buffered_output.append(
+                        f"{rule_instance.rule_config.qualname}: [yellow]no tests[/yellow] in {rule_instance.rule_config.test_path}"
+                    )
                 else:
-                    key = progress.add_task(rule_instance.rule_config.qualname, total=len(names))
+                    key = str(i)
+                    i += 1
                     for n in names:
-                        outstanding[tpe.submit(self._perform_test, rule_instance, n)] = (key, rule_instance.rule_config.qualname)
-
-            progress.update(prepare_key, completed=True)
-            # breakpoint()
-
-            for fut in as_completed(outstanding.keys()):
-                progress_key, desc = outstanding[fut]
-                try:
-                    fut.result()
-                except Exception as e:
-                    progress.console.print(desc)
-                    progress.console.print("  " + repr(e))
+                        outstanding[tpe.submit(self._perform_test, rule_instance, n)] = (
+                            key,
+                            f"{rule_instance.rule_config.qualname} on {n}",
+                        )
+
+                success = True
+                for fut in outstanding.keys():
+                    progress_key, desc = outstanding[fut]
+                    try:
+                        fut.result()
+                    except Exception as e:
+                        print("[red]F[/red]", end="")
+                        buffered_output.append(f"  {desc}:")
+                        # This should be combined with how we actually run
+                        # things...
+                        typ, value, tb = sys.exc_info()
+                        buf = io.StringIO()
+                        traceback.print_tb(tb, file=buf)
+                        # TODO redent
+                        buffered_output.append("  " + buf.getvalue())
+                        buffered_output.append(repr(e))
+                        success = False
+                    else:
+                        print(".", end="")
+
+                # TODO try to line these up
```

Comment:
If you mean across rules, we could determine the max number of tests before running any.

Reply:
Yes, definitely puntable.